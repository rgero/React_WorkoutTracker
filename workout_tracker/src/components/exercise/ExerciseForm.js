import React from 'react';
import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container';
import Form from 'react-bootstrap/Form';
import InputGroup from 'react-bootstrap/InputGroup';

import { SetForm } from '../set/SetForm';
import { SetList } from '../set/SetList';

export const ExerciseForm = ({exercise = {}, onSubmit})=> 
{
    const [setList, setSetList] = React.useState(exercise.setList ? exercise.setList : []);
    const [muscleGroup, setMuscleGroup] = React.useState(exercise.muscleGroup ? exercise.muscleGroup : "");
    const [name, setExerciseName] = React.useState(exercise.name ? exercise.name : "");
    const [notes, setNotes] = React.useState(exercise.notes ? exercise.notes : "");
    const [errMsg, setError] = React.useState("");

    const addSet = (set) => {
        setSetList([...setList, set]);
    }
    
    const deleteSet = (index) => {
        let newSetList = setList;
        newSetList.splice(index, 1);
        setSetList( [...newSetList] );
    }

    const processExercise = (e) => {
        e.preventDefault();

        if (name === "")
        {
            setError("Missing exercise name");
            return;
        }

        if (setList.length === 0)
        {
            setError("Missing set list");
            return;
        }

        const newExercise = {
            name,
            muscleGroup,
            setList,
            notes
        };

        onSubmit(newExercise);
        setSetList([]);
        setMuscleGroup("")
        setExerciseName("");
        setNotes("");
        setError("");
    }

    return (
        <Container fluid="md">
            <Form id="addExerciseForm" onSubmit={processExercise}>
                <InputGroup className="mb-3">
                    <InputGroup.Text>Name</InputGroup.Text>
                    <Form.Control
                        aria-label="name" 
                        type="text" 
                        value={name} 
                        onChange={e => setExerciseName(e.target.value)}
                    />
                </InputGroup>
                <InputGroup className="mb-3">
                    <InputGroup.Text>Muscle Group</InputGroup.Text>
                    <Form.Control
                        aria-label="muscleGroup"
                        type="text"
                        value={muscleGroup}
                        onChange={e => setMuscleGroup(e.target.value)}
                    />
                </InputGroup>
                <InputGroup className="mb-3">
                    <InputGroup.Text>Notes</InputGroup.Text>
                    <Form.Control
                        aria-label="exerciseNotes"
                        as="textarea"
                        value={notes}
                        rows={3}
                        onChange={e => setNotes(e.target.value)}
                    />
                </InputGroup>
            </Form>
            <SetForm onSubmit={addSet}/>
            <SetList setList={setList} onDelete={deleteSet}/>
            <Button
                aria-label="exerciseSubmit" 
                className="mt-4" 
                form="addExerciseForm" 
                variant="outline-secondary" 
                type="submit"
            >
                    Add Exercise
            </Button>

            {/* Error Processing? Don't know if this is completely necessary */}
            { errMsg ? (
                <Container className="pt-4">
                    <Alert key='danger' variant='danger'>
                        Error: {errMsg}
                    </Alert>
                </Container>
            ): ( null )}
        </Container>
    )
}