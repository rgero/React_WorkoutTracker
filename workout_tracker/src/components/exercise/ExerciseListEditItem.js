import React from 'react';

import { SetList } from '../set/SetList';
import {AlertBox} from '../../helpers/DialogBox';

import Container from 'react-bootstrap/Container';
import Card from 'react-bootstrap/Card';

export const ExerciseListItem = ({index, exercise = {}, onDelete})=> 
{
  const processDelete = () => {
    let title = 'Are you sure you want to delete this?';
    let subtitle = `The exercise named ${exercise.name}`;
    let buttons = [
      {
        label: 'Yes',
        onClick: async () => {
          await onDelete(exercise);
        }
      },
      {
        label: 'No'
      }
    ]
    AlertBox(title, subtitle, buttons);
  }

  return (
    <>
      {
        Object.keys(exercise).length === 0 ? (
          <Container className="list-item list-item--message">
            <span>No Name</span>
          </Container>
        ) : (
          <Card onClick={processDelete}>
            <Card.Body>
              <Card.Title>#{index} - {exercise.name}</Card.Title>
              <Card.Body>
                <SetList  setList={exercise.setList} key={`${exercise.name}_Sets`}/>
              </Card.Body>
            </Card.Body>
          </Card>
        )
      }
    </>
  )
}