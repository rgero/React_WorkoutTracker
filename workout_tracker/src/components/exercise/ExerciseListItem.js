import React from 'react';

import { SetList } from '../set/SetList';
import {AlertBox} from '../../helpers/DialogBox';

import Container from 'react-bootstrap/Container';

export const ExerciseListItem = ({exercise, onDelete})=> 
{
  let exerciseName = exercise.name;
  let setList = exercise.setList;

  const processDelete = () => {
    let title = 'Are you sure you want to delete this?';
    let subtitle = `The exercise named ${exerciseName}`;
    let buttons = [
      {
        label: 'Yes',
        onClick: async () => {
          await onDelete(exercise);
        }
      },
      {
        label: 'No'
      }
    ]
    AlertBox(title, subtitle, buttons);
  }

  return (
    <>
      {
        exerciseName === "" ? (
          <Container className="list-item list-item--message">
            <span>No Name</span>
          </Container>
        ) : (
            <Container>
                <Container onClick={processDelete}>
                    {exerciseName}
                    <SetList setList={setList} key={`${exerciseName}_Sets`}/>
                </Container>
            </Container>
        )
      }
    </>
  )
}