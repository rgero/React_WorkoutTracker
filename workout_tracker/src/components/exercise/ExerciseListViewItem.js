import React from 'react';

import { SetList } from '../set/SetList';
import {AlertBox} from '../../helpers/DialogBox';

import Container from 'react-bootstrap/Container';
import Card from 'react-bootstrap/Card';

export const ExerciseListItem = ({index, exercise = {}})=> 
{

  return (
    <>
      {
        Object.keys(exercise).length === 0 ? (
          <Container className="list-item list-item--message">
            <span>No Name</span>
          </Container>
        ) : (
          <Card>
            <Card.Body>
              <Card.Title>#{index} - {exercise.name}</Card.Title>
              <Card.Body>
                <SetList  setList={exercise.setList} key={`${exercise.name}_Sets`}/>
              </Card.Body>
            </Card.Body>
          </Card>
        )
      }
    </>
  )
}