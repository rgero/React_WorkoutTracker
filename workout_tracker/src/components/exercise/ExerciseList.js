import React from 'react';
import Container from 'react-bootstrap/Container';
import ListGroup from 'react-bootstrap/ListGroup';

import { ExerciseListItem } from '../exercise/ExerciseListItem';

export const ExerciseList = ({exerciseList = [], onDelete})=> 
{
    return (
        <Container>
            {
                exerciseList.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No Exercises</span>
                    </div>
                ) : (
                    <Container>
                        <ListGroup>
                        {
                            exerciseList.map((exercise, index) => (
                                <ListGroup.Item action onClick={(e) => onDelete(index)}>
                                    <ExerciseListItem
                                        exercise={exercise}
                                    />
                                </ListGroup.Item>
                            ))
                        }
                        </ListGroup>

                    </Container>
                )
            }
        </Container>
    )
}