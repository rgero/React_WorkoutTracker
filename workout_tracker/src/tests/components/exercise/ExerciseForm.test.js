import TestRenderer from 'react-test-renderer';

import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event'
import { act } from 'react-dom/test-utils';

import {ExerciseForm} from '../../../components/exercise/ExerciseForm';
import {TestExercise} from '../../testData';

describe("ExerciseForm Rendering Tests", () => {
    test('No data', ()=> {
        let element = new TestRenderer.create(
            <ExerciseForm onSubmit={jest.fn()} />
        )
        expect(element.toJSON()).toMatchSnapshot();
    })

    test('With Data', ()=> {
        let element = new TestRenderer.create(
            <ExerciseForm
                exercise = {TestExercise}
                onSubmit={jest.fn()} 
            />
        )
        expect(element.toJSON()).toMatchSnapshot();
    })
})

describe("ExerciseForm Submissions", ()=> {

    let nameInput, muscleInput, notesInput;
    let exerciseSubmit;
    
    let user;

    let mockSubmit = jest.fn();

    let dummyExerciseWithSet = {
        "setList": [
            {
                "reps": 40,
                "weight": 100,
            }
        ]
    }

    beforeEach(()=> {
        user = userEvent.setup();
        render(<ExerciseForm exercise={dummyExerciseWithSet} onSubmit={mockSubmit} />)

        act(()=> {
            nameInput = screen.getByRole("textbox", {name: /name/i});
            muscleInput = screen.getByRole("textbox", {name: /musclegroup/i});
            notesInput = screen.getByRole("textbox", {name: /exerciseNotes/i});

            exerciseSubmit = screen.getByRole("button", {name: /exercisesubmit/i});
        })

    })

    test("Minimum Valid Submit", async ()=> {
        await user.click(nameInput);
        await user.keyboard(TestExercise.name);
        await user.click(exerciseSubmit);

        let expectedResult = {
            name: TestExercise.name,
            notes: "",
            muscleGroup: "",
            setList: dummyExerciseWithSet.setList
        }

        expect(mockSubmit).toHaveBeenCalled();
        expect(mockSubmit).toHaveBeenCalledWith(expectedResult)
    })

    test("Adding a note, valid submit", async ()=> {

        await user.click(nameInput);
        await user.keyboard(TestExercise.name);
        await user.click(notesInput);
        await user.keyboard(TestExercise.notes);
        await user.click(exerciseSubmit);

        let expectedResult = {
            name: TestExercise.name,
            notes: TestExercise.notes,
            muscleGroup: "",
            setList: dummyExerciseWithSet.setList
        }

        expect(mockSubmit).toHaveBeenCalledWith(expectedResult)
    })

    test("Adding a muscle group, valid submit", async ()=> {
        await user.click(nameInput);
        await user.keyboard(TestExercise.name);

        await user.click(muscleInput);
        await user.keyboard(TestExercise.muscleGroup);
        
        await user.click(exerciseSubmit);

        let expectedResult = {
            name: TestExercise.name,
            notes: "",
            muscleGroup: TestExercise.muscleGroup,
            setList: dummyExerciseWithSet.setList
        }

        expect(mockSubmit).toHaveBeenCalledWith(expectedResult)
    })

    test("Full Form, valid submit", async ()=> {
        await user.click(nameInput);
        await user.keyboard(TestExercise.name);

        await user.click(notesInput);
        await user.keyboard(TestExercise.notes);

        await user.click(muscleInput);
        await user.keyboard(TestExercise.muscleGroup);
        
        await user.click(exerciseSubmit);

        let expectedResult = {
            name: TestExercise.name,
            notes: TestExercise.notes,
            muscleGroup: TestExercise.muscleGroup,
            setList: dummyExerciseWithSet.setList
        }

        expect(mockSubmit).toHaveBeenCalled();
        expect(mockSubmit).toHaveBeenCalledWith(expectedResult)
    })
})

describe("Form Invalid Submissions", ()=> {

    let mockSubmit = jest.fn();
    let dummyExerciseWithSet = {
        setList: TestExercise.setList
    }

    let user = userEvent.setup();
    

    test("Missing Name", async ()=> {
        render(<ExerciseForm exercise={dummyExerciseWithSet} onSubmit={mockSubmit} />)

        let exerciseSubmit = screen.getByRole("button", {name: /exercisesubmit/i});
        await user.click(exerciseSubmit);

        expect(mockSubmit).not.toHaveBeenCalled();

        const errorField = screen.getByText("Error: Missing exercise name");
        expect(errorField).toBeInTheDocument();
    })

    test("Missing set list", async ()=> {

        render(<ExerciseForm onSubmit={mockSubmit} />)

        let nameInput = screen.getByRole("textbox", {name: /name/i});
        await user.click(nameInput);
        await user.keyboard(TestExercise.name);
            
        let exerciseSubmit = screen.getByRole("button", {name: /exercisesubmit/i});
        await user.click(exerciseSubmit);

        expect(mockSubmit).not.toHaveBeenCalled();

        const errorField = screen.getByText("Error: Missing set list");
        expect(errorField).toBeInTheDocument();

    })

})