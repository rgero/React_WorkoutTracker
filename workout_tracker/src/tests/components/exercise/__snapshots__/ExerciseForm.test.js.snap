// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseForm Rendering Tests No data 1`] = `
<div
  className="container-md"
>
  <div
    className="row"
  >
    <form
      className=""
      id="addExerciseForm"
      onSubmit={[Function]}
    >
      <div
        className="mb-3 form-floating"
      >
        <input
          aria-label="name"
          className="form-control"
          id="floatingExerciseName"
          onChange={[Function]}
          type="text"
          value=""
        />
        <label
          htmlFor="floatingExerciseName"
        >
          Name
        </label>
      </div>
      <div
        className="mb-3 form-floating"
      >
        <input
          aria-label="muscleGroup"
          className="form-control"
          id="floatingMuscleGroup"
          onChange={[Function]}
          type="text"
          value=""
        />
        <label
          htmlFor="floatingMuscleGroup"
        >
          Muscle Group
        </label>
      </div>
      <div
        className="mb-3 form-floating"
      >
        <textarea
          aria-label="exerciseNotes"
          className="form-control"
          id="floatingExerciseNotes"
          onChange={[Function]}
          rows={3}
          value=""
        />
        <label
          htmlFor="floatingExerciseNotes"
        >
          Notes
        </label>
      </div>
    </form>
  </div>
  <div
    className="row"
  >
    <div
      className="container-md"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="justify-content-md-center align-items-md-center row row-cols-md-4"
        >
          <div
            className="col-md"
          >
            <div
              className="mb-3 form-floating"
            >
              <input
                aria-label="reps"
                className="form-control"
                id="floatingReps"
                onChange={[Function]}
                type="text"
                value=""
              />
              <label
                htmlFor="floatingReps"
              >
                Reps
              </label>
            </div>
          </div>
          <div
            className="col-md"
          >
            <div
              className="mb-3 form-floating"
            >
              <input
                aria-label="weight"
                className="form-control"
                id="floatingWeight"
                onChange={[Function]}
                type="text"
                value=""
              />
              <label
                htmlFor="floatingWeight"
              >
                Weight
              </label>
            </div>
          </div>
        </div>
        <button
          aria-label="setSubmit"
          className="float-end btn btn-outline-secondary"
          disabled={false}
          type="submit"
        >
          Add Set
        </button>
      </form>
    </div>
  </div>
  <div
    className="row"
  >
    <button
      aria-label="exerciseSubmit"
      className="mt-4 btn btn-outline-secondary"
      disabled={false}
      form="addExerciseForm"
      type="submit"
    >
      Add Exercise
    </button>
  </div>
</div>
`;

exports[`ExerciseForm Rendering Tests With Data 1`] = `
<div
  className="container-md"
>
  <div
    className="row"
  >
    <form
      className=""
      id="addExerciseForm"
      onSubmit={[Function]}
    >
      <div
        className="mb-3 form-floating"
      >
        <input
          aria-label="name"
          className="form-control"
          id="floatingExerciseName"
          onChange={[Function]}
          type="text"
          value="Lunges"
        />
        <label
          htmlFor="floatingExerciseName"
        >
          Name
        </label>
      </div>
      <div
        className="mb-3 form-floating"
      >
        <input
          aria-label="muscleGroup"
          className="form-control"
          id="floatingMuscleGroup"
          onChange={[Function]}
          type="text"
          value="Legs"
        />
        <label
          htmlFor="floatingMuscleGroup"
        >
          Muscle Group
        </label>
      </div>
      <div
        className="mb-3 form-floating"
      >
        <textarea
          aria-label="exerciseNotes"
          className="form-control"
          id="floatingExerciseNotes"
          onChange={[Function]}
          rows={3}
          value="Dumbbells"
        />
        <label
          htmlFor="floatingExerciseNotes"
        >
          Notes
        </label>
      </div>
    </form>
  </div>
  <div
    className="row"
  >
    <div
      className="container-md"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="justify-content-md-center align-items-md-center row row-cols-md-4"
        >
          <div
            className="col-md"
          >
            <div
              className="mb-3 form-floating"
            >
              <input
                aria-label="reps"
                className="form-control"
                id="floatingReps"
                onChange={[Function]}
                type="text"
                value=""
              />
              <label
                htmlFor="floatingReps"
              >
                Reps
              </label>
            </div>
          </div>
          <div
            className="col-md"
          >
            <div
              className="mb-3 form-floating"
            >
              <input
                aria-label="weight"
                className="form-control"
                id="floatingWeight"
                onChange={[Function]}
                type="text"
                value=""
              />
              <label
                htmlFor="floatingWeight"
              >
                Weight
              </label>
            </div>
          </div>
        </div>
        <button
          aria-label="setSubmit"
          className="float-end btn btn-outline-secondary"
          disabled={false}
          type="submit"
        >
          Add Set
        </button>
      </form>
    </div>
    <div
      className="pt-3 container"
      style={
        Object {
          "margin": "auto",
          "textAlign": "center",
          "width": "60%",
        }
      }
    >
      <div
        style={
          Object {
            "textAlign": "left",
          }
        }
      >
        Set List
      </div>
      <table
        className="table table-sm table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Reps
            </th>
            <th>
              Weight
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            onClick={[Function]}
          >
            <td>
              40
            </td>
            <td>
              100
            </td>
          </tr>
          <tr
            onClick={[Function]}
          >
            <td>
              40
            </td>
            <td>
              100
            </td>
          </tr>
          <tr
            onClick={[Function]}
          >
            <td>
              40
            </td>
            <td>
              100
            </td>
          </tr>
          <tr
            onClick={[Function]}
          >
            <td>
              40
            </td>
            <td>
              100
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="row"
  >
    <button
      aria-label="exerciseSubmit"
      className="mt-4 btn btn-outline-secondary"
      disabled={false}
      form="addExerciseForm"
      type="submit"
    >
      Add Exercise
    </button>
  </div>
</div>
`;
